### PR TITLE
Fix race binlog

### PR DESCRIPTION
**Pull request by s-mx**

### CODE REVIEW

File: library/blob.c [hunk 1]
```diff
@@ -2028,11 +2025,32 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
 			copy, copy_offset, old);
 
 err_out_exit:
+	return err;
+}
+
+
+/**
+ * eblob_write_prepare_disk() - allocates space for new record
```
Comment (s-mx) [inline library/blob.c:2035]: fix the description

File: library/blob.c [hunk 2]
```diff
@@ -2126,6 +2150,9 @@ int eblob_write_prepare(struct eblob_backend *b, struct eblob_key *key,
 err_out_cleanup_wc:
 	eblob_write_control_cleanup(&wc);
 err_out_exit:
+	if (!write_to_base)
```
Comment (s-mx) [inline library/blob.c]: maybe bad way to unlock backend

File: library/blob.c [hunk 3]
```diff
@@ -2028,11 +2025,32 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
+ * It allocates new bases, commits headers and manages overwrites/appends.
+ */
+static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *key,
+				    struct eblob_write_control *wc, uint64_t prepare_disk_size,
+				    enum eblob_copy_flavour copy, uint64_t copy_offset, struct eblob_ram_control *old,
+				    size_t defrag_generation)
+{
+	ssize_t err = 0;
+	eblob_log(b->cfg.log, EBLOB_LOG_NOTICE,
+		  "blob: %s: eblob_write_prepare_disk: start: "
```
Comment (antmat) [inline library/blob.c]: looks like exceeded logging

File: library/blob.c [hunk 4]
```diff
@@ -2063,11 +2082,16 @@ int eblob_write_prepare(struct eblob_backend *b, struct eblob_key *key,
 	defrag_generation = b->defrag_generation;
 
 	err = eblob_fill_write_control_from_ram(b, key, &wc, 1, &old);
-	pthread_mutex_unlock(&b->lock);
+	if (err == 0 && (wc.total_size >= eblob_calculate_size(b, key, 0, size)) &&
```
Comment (antmat) [inline library/blob.c]: make some comments about possible cases and locking strategy

**Review by antmat — CHANGES_REQUESTED** [review]
Looks like it's correct, however there are minor notices and someone else should also take a look on this PR

File: library/blob.c [hunk 5]
```diff
@@ -1986,23 +1986,20 @@ static int eblob_write_prepare_disk_ll(struct eblob_backend *b, struct eblob_key
...
 
 	eblob_log(b->cfg.log, EBLOB_LOG_NOTICE,
-			"blob: %s: eblob_write_prepare_disk: start: "
-			"size: %" PRIu64 ", offset: %" PRIu64 ", prepare: %" PRIu64 "\n",
-			eblob_dump_id(key->id), wc->size, wc->offset, prepare_disk_size);
-
-	pthread_mutex_lock(&b->lock);
+		  "blob: %s: eblob_write_prepare_disk: start: "
+		  "size: %" PRIu64 ", offset: %" PRIu64 ", prepare: %" PRIu64 "\n",
+		eblob_dump_id(key->id), wc->size, wc->offset, prepare_disk_size);
```
Comment (shaitan) [inline library/blob.c]: There is no changes connected with PR. Please return it to original.

File: library/blob.c [hunk 6]
```diff
@@ -2028,11 +2025,32 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
+static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *key,
+				    struct eblob_write_control *wc, uint64_t prepare_disk_size,
+				    enum eblob_copy_flavour copy, uint64_t copy_offset, struct eblob_ram_control *old,
+				    size_t defrag_generation)
+{
+	ssize_t err = 0;
+	eblob_log(b->cfg.log, EBLOB_LOG_NOTICE,
+		  "blob: %s: eblob_write_prepare_disk: start: "
+		  "size: %" PRIu64 ", offset: %" PRIu64 ", prepare: %" PRIu64 "\n",
+		eblob_dump_id(key->id), wc->size, wc->offset, prepare_disk_size);
```
Comment (sabramkin) [inline library/blob.c]: Some is wrong with indentation, column is violated.

File: library/blob.c (hunk 2 above)
Comment (sabramkin) [reply library/blob.c]: The code is congruent enough: either (write_to_base == 0, no lock) or (write_to_base == 1, lock). Seems ok independently on any label jumps.

File: library/blob.c [hunk 7]
```diff
@@ -2028,11 +2025,32 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
+	return err;
+}
+
+
+/**
+ * eblob_write_prepare_disk() - allocates space for new record
+ * It allocates new bases, commits headers and manages overwrites/appends.
+ */
+static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *key,
+				    struct eblob_write_control *wc, uint64_t prepare_disk_size,
```
Comment (shaitan) [inline library/blob.c]:
```suggestion
                                    struct eblob_write_control *wc, uint64_t prepare_disk_size,
```

File: library/blob.c (hunk 1 above)
Comment (s-mx) [reply library/blob.c:2035]: done

File: library/blob.c (hunk 2 above)
Comment (s-mx) [reply library/blob.c]: now it's redone.

File: library/blob.c (hunk 3 above)
Comment (s-mx) [reply library/blob.c]: fixed

File: library/blob.c (hunk 5 above)
Comment (s-mx) [reply library/blob.c]: fixed

File: library/blob.c (hunk 4 above)
Comment (s-mx) [reply library/blob.c]: done

File: library/blob.c [hunk 8]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
-				goto err_out_cleanup_wc;
-
-			err = eblob_commit_ram(b, key, &wc);
-			if (err)
-				goto err_out_cleanup_wc;
-		}
-
-		eblob_stat_inc(b->stat, EBLOB_GST_PREPARE_REUSED);
-		goto err_out_cleanup_wc;
+	// Note: were we have 4 cases:
```
Comment (antmat) [inline library/blob.c]:
```suggestion
	// Note: were we have 3 cases:
```

**Review by antmat — CHANGES_REQUESTED** [review]

File: library/blob.c (hunk 8 above)
Comment (s-mx) [reply library/blob.c]: done.

File: library/blob.c (hunk 6 above)
Comment (s-mx) [reply library/blob.c]: @sabramkin check it now, please.

File: library/blob.c [hunk 9]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
-				goto err_out_cleanup_wc;
-
-			err = eblob_commit_ram(b, key, &wc);
-			if (err)
-				goto err_out_cleanup_wc;
-		}
-
-		eblob_stat_inc(b->stat, EBLOB_GST_PREPARE_REUSED);
-		goto err_out_cleanup_wc;
+	// Note: were we have 3 cases:
```
Comment (shaitan) [inline library/blob.c]: `were`?

File: library/blob.c [hunk 10]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
+static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *key,
+		struct eblob_write_control *wc, uint64_t prepare_disk_size,
+		enum eblob_copy_flavour copy, uint64_t copy_offset, struct eblob_ram_control *old,
+		size_t defrag_generation)
+{
+	pthread_mutex_lock(&b->lock);
+	int err = eblob_write_prepare_disk_nolock(b, key, wc, prepare_disk_size,
+	                                          copy, copy_offset, old, defrag_generation);
 	pthread_mutex_unlock(&b->lock);
-	eblob_dump_wc(b, key, wc, "eblob_write_prepare_disk", err);
```
Comment (shaitan) [inline library/blob.c]: This log was lost

File: library/blob.c [hunk 11]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
-		err = eblob_write_prepare_disk(b, key, &wc, size, EBLOB_COPY_RECORD, 0, err == -ENOENT ? NULL : &old, defrag_generation);
+		err = eblob_write_prepare_disk_nolock(b, key, &wc, size, EBLOB_COPY_RECORD, 0,
+		                                      err == -ENOENT ? NULL : &old, defrag_generation);
+		pthread_mutex_unlock(&b->lock);
 		if (err)
 			goto err_out_cleanup_wc;
 
 		err = eblob_commit_ram(b, key, &wc);
-		if (err)
-			goto err_out_cleanup_wc;
```
Comment (shaitan) [inline library/blob.c]: Why was it removed?

File: library/blob.c [hunk 12]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
 	pthread_mutex_unlock(&b->lock);
-	eblob_dump_wc(b, key, wc, "eblob_write_prepare_disk", err);
 	return err;
 }
 
+
+/**
+ * write_prepare_inplace() - prepare space in existing base. Non-blocking.
+ */
+static int write_prepare_inplace(struct eblob_backend *b, struct eblob_key *key,
```
Comment (shaitan) [inline library/blob.c]:
```suggestion
static int eblob_write_prepare_inplace(struct eblob_backend *b, struct eblob_key *key,
```

File: library/blob.c [hunk 13]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
+		uint64_t flags, struct eblob_write_control *wc) {
+	// We've got a key which will be overwritten, make sure it has valid flags.
+	// We overwrite flags to what user has provided dropping all existing on-disk flags since
+	// given key will be fully overwritten and we do not care about its old content anymore.
+	// The same logic is performed in @eblob_write_prepare_disk(), but it also allocates new space.
+
+	int err = 0;
+	uint64_t new_flags = 0;
```
Comment (shaitan) [inline library/blob.c]:
```suggestion
	uint64_t new_flags = eblob_validate_ctl_flags(b, flags);
```

File: library/blob.c [hunk 14]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
+	new_flags = eblob_validate_ctl_flags(b, flags);
+	new_flags |= BLOB_DISK_CTL_UNCOMMITTED;
+
+	if (wc->flags == new_flags) {
+		goto out_stat;
```
Comment (shaitan) [inline library/blob.c:2067]: Why isn't it just:
```c
eblob_stat_inc(b->stat, EBLOB_GST_PREPARE_REUSED);
goto out_exit;
```

File: library/blob.c [hunk 15]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
+}
+
+
 /**
  * eblob_write_prepare() - prepare phase reserves space in blob file.
  */
 int eblob_write_prepare(struct eblob_backend *b, struct eblob_key *key,
 		uint64_t size, uint64_t flags)
 {
-	FORMATTED(HANDY_TIMER_SCOPE, ("eblob.%u.disk.write.prepare", b->cfg.stat_id));
```
Comment (shaitan) [inline library/blob.c]: Why was it moved?

File: library/blob.c [hunk 16]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
 	struct eblob_write_control wc = { .offset = 0 };
 	struct eblob_ram_control old;
 	int err;
 	size_t defrag_generation = 0;
 
-	EBLOB_WARNX(b->cfg.log, EBLOB_LOG_DEBUG,
```
Comment (shaitan) [inline library/blob.c]: Why was it moved?

File: library/blob.c [hunk 17]
```diff
@@ -2028,99 +2026,128 @@ static int eblob_write_prepare_disk(struct eblob_backend *b, struct eblob_key *k
...
-		goto err_out_cleanup_wc;
+	// Note: were we have 3 cases:
+	// 1) We have not found a key. => hold backend lock, reserve space in the new base
+	// 2) We've found a key in existing bases.
+	//   2.a) Corresponding base has no enough space. => hold backend lock, reserve in the new base
+	//   2.b) Corresponding base has closed for writing (binlog enabled) => release backend lock,
+	//                                                                      reserve in the new base
+	if (err == 0 && (wc.total_size >= eblob_calculate_size(b, key, 0, size)) &&
+	    !eblob_binlog_enabled(&wc.bctl->binlog)) {
+		pthread_mutex_unlock(&b->lock);
```
Comment (shaitan) [inline library/blob.c:2139]: Shouldn't it be unlocked after `eblob_commit_disk`?

File: library/blob.c (hunk 8 above)
Comment (sabramkin) [reply library/blob.c]: Minute of humour: `3` is hardcode, `some cases` or `a couple of cases` would be more flexible and supportable =)

File: library/blob.c (hunk 14 above)
Comment (s-mx) [reply library/blob.c:2067]: we want to inc stat in another success cases

File: library/blob.c (hunk 16 above)
Comment (s-mx) [reply library/blob.c]: we dereference b and after that check that it's not NULL. stupid.

File: library/blob.c (hunk 17 above)
Comment (s-mx) [reply library/blob.c:2139]: It seem's that bctl critness equal 1 is enough.

File: library/blob.c (hunk 14 above)
Comment (shaitan) [reply library/blob.c:2067]: You are right, thanks

File: library/blob.c (hunk 9 above)
Comment (s-mx) [reply library/blob.c]: here

File: library/blob.c (hunk 16 above)
Comment (shaitan) [reply library/blob.c]: Could you please move this change into a separate commit?

File: library/blob.c (hunk 17 above)
Comment (shaitan) [reply library/blob.c:2139]: ok

File: library/blob.c (hunk 15 above)
Comment (s-mx) [reply library/blob.c]: returned

File: library/blob.c (hunk 16 above)
Comment (s-mx) [reply library/blob.c]: returned

File: library/blob.c (hunk 13 above)
Comment (s-mx) [reply library/blob.c]: done

File: library/blob.c (hunk 12 above)
Comment (s-mx) [reply library/blob.c]: done

File: library/blob.c (hunk 10 above)
Comment (s-mx) [reply library/blob.c]: done

File: library/blob.c (hunk 9 above)
Comment (s-mx) [reply library/blob.c]: fixed

File: library/blob.c (hunk 11 above)
Comment (s-mx) [reply library/blob.c]: because there is nothing between this line and that label.
returned it back.

**Review by antmat — DISMISSED** [review]

**Comment by s-mx** [comment]
Additionally tests are needed.

**Review by antmat — DISMISSED** [review]
I'd think a little about extracting common part from tests, but it looks good to me already.

File: library/datasort.c [hunk 18]
```diff
@@ -1128,6 +1128,108 @@ static void datasort_destroy(struct datasort_ctl *ds_ctl)
 	ds_ctl->chunks_dir = NULL;
 }
 
+/* Binlog entry mgmt subroutines */
+
+__attribute__ ((warn_unused_result))
+__attribute__ ((nonnull))
+inline struct eblob_binlog_entry *
+eblob_binlog_entry_new(const struct eblob_key *key)
```
Comment (shaitan) [inline library/datasort.c]: Why is it needed? How is it connected with this commit?

File: tests/CMakeLists.txt [hunk 19]
```diff
@@ -31,11 +31,19 @@ add_custom_target(test_defrag
                   COMMAND "${CMAKE_CURRENT_BINARY_DIR}/eblob_defrag_test"
                   DEPENDS ${TESTS_DEPS} eblob_defrag_test)
 
+
+add_executable(eblob_binlog_test unit/binlog_test.cpp unit/eblob_wrapper.cpp)
```
Comment (shaitan) [inline tests/CMakeLists.txt]: It should be in the next commit with tests

File: library/blob.c [hunk 20]
```diff
@@ -2049,10 +2110,11 @@ int eblob_write_prepare(struct eblob_backend *b, struct eblob_key *key,
 			"key: %s, size: %" PRIu64 ", flags: %s",
 			eblob_dump_id(key->id), size, eblob_dump_dctl_flags(flags));
 
+
 	/* Sanity */
 	if (b == NULL || key == NULL) {
 		err = -EINVAL;
-		goto err_out_exit;
+		return err;
```
Comment (shaitan) [inline library/blob.c]: This change drops log with `eblob_write_prepare: finished`. Why?

File: library/datasort.c (hunk 18 above)
Comment (s-mx) [reply library/datasort.c]: returned it back

File: tests/unit/binlog_test.cpp [hunk 21]
```diff
@@ -0,0 +1,117 @@
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MODULE DEFRAG library test
```
Comment (shaitan) [inline tests/unit/binlog_test.cpp]: `DEFRAG`?

File: tests/unit/binlog_test.cpp [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+
+#include <boost/test/included/unit_test.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include "eblob_wrapper.h"
+
+#include "library/datasort.h"
+
+#include <vector>
```
Comment (shaitan) [inline tests/unit/binlog_test.cpp]: global includes should be placed before local

File: tests/unit/binlog_test.cpp [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+	eblob_config &config = config_wrapper.config;
+	constexpr size_t RECORDS_IN_BLOB = 10;
+	config.records_in_blob = RECORDS_IN_BLOB;
+	return config_wrapper;
+}
+
+/**
+ * Make a blob with 10 records 100 length each.
+ * Invoke write_prepare with key in existing base and smaller size.
+ * Check that number of bases didn't change
```
Comment (shaitan) [inline tests/unit/binlog_test.cpp]:
```suggestion
 * Check that number of bases wasn't changed
```

File: tests/unit/binlog_test.cpp [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+	eblob_wrapper wrapper(config);
+	std::vector<item_t> shadow_elems;
+	auto generator = make_uniform_item_generator(wrapper, 100, 100);
+	BOOST_REQUIRE_EQUAL(fill_eblob(wrapper, shadow_elems, generator, TOTAL_RECORDS), 0);
+	BOOST_REQUIRE_EQUAL(wrapper.number_bases(), 1);
+	eblob_base_ctl *bctl = container_of(wrapper.get()->bases.next, eblob_base_ctl, base_entry);
+	BOOST_REQUIRE_EQUAL(eblob_binlog_start(&bctl->binlog), 0);
+	BOOST_REQUIRE_EQUAL(eblob_write_prepare(wrapper.get(), &shadow_elems.back().hashed_key, 50, 0), 0);
+	BOOST_REQUIRE_EQUAL(wrapper.number_bases(), 2);
+}
```
Comment (shaitan) [inline tests/unit/binlog_test.cpp]: Empty line is needed at the end of each file

File: library/datasort.h [hunk 25]
```diff
@@ -183,7 +183,7 @@ static void eblob_binlog_entry_free(struct eblob_binlog_entry *entry)
 __attribute__ ((nonnull (1)))
 static inline struct eblob_binlog_entry *
 eblob_binlog_iterate(const struct eblob_binlog_cfg *bcfg,
-		const struct eblob_binlog_entry *it)
+		     const struct eblob_binlog_entry *it)
```
Comment (shaitan) [inline library/datasort.h]: Please, return whole file to original state.

File: tests/unit/binlog_test.cpp (hunk 23 above)
Comment (s-mx) [reply tests/unit/binlog_test.cpp]: fixed

File: library/datasort.h (hunk 25 above)
Comment (s-mx) [reply library/datasort.h]: returned

File: tests/unit/binlog_test.cpp (hunk 21 above)
Comment (s-mx) [reply tests/unit/binlog_test.cpp]: fixed

File: tests/CMakeLists.txt (hunk 19 above)
Comment (s-mx) [reply tests/CMakeLists.txt]: fixed

File: library/datasort.h [hunk 26]
```diff
@@ -248,6 +248,7 @@ static inline int eblob_binlog_append(struct eblob_binlog_cfg *bcfg,
 	return 0;
 }
 
+
```
Comment (shaitan) [inline library/datasort.h]: The last one)

File: library/blob.c (hunk 20 above)
Comment (s-mx) [reply library/blob.c]: returned

File: tests/unit/binlog_test.cpp (hunk 22 above)
Comment (s-mx) [reply tests/unit/binlog_test.cpp]: fixed

File: library/datasort.h (hunk 26 above)
Comment (s-mx) [reply library/datasort.h]: fixed

File: tests/unit/binlog_test.cpp (hunk 24 above)
Comment (s-mx) [reply tests/unit/binlog_test.cpp]: fixed here

File: tests/unit/binlog_test.cpp [hunk 27]
```diff
@@ -0,0 +1,132 @@
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MODULE BINLOG library test
+
+#include <boost/test/included/unit_test.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include "eblob_wrapper.h"
+
+#include <library/datasort.h>
```
Comment (shaitan) [inline tests/unit/binlog_test.cpp]: It is a local include, so it should be in `""`

Comment (s-mx) [reply tests/unit/binlog_test.cpp]: done

**Review by shaitan — APPROVED** [review]
LGTM, thanks

**Pull request merged**